### ISSUE TITLE
Cannot compile for ESP8266 using Docker image, linking step fails

### ISSUE THREAD

**Issue by djryanj** [esphome/issues#2878]
### The problem

I have been unable to compile firmware for an ESP8266 board, the linking step fails consistently with: `xtensa-lx106-elf-g++: fatal error: -fuse-linker-plugin, but liblto_plugin.so not found`

### Which version of ESPHome has the issue?

Docker tags 2021.12.1, 2022.1.0-dev, 2021.11.1, 2021.10.3

### What type of installation are you using?

Docker

### What platform are you using?

ESP8266

### Board

Wemos D1 Mini, also esp01_1m

### Component causing the issue

Unknown

### Example YAML snippet

```yaml
esphome:
  name: outdoorplug

esp8266:
  board: d1_mini

wifi:
  ssid: "ssid"
  password: "password"
  ap:
    ssid: "d1-hotspot"
    password: "password"

captive_portal:

logger:

ota:
  password: ""

switch:
  - platform: gpio
    name: "Plug"
    pin: D6
    id: relay

status_led:
  pin:
    number: D4
    inverted: true

mqtt:
  topic_prefix: outdoors/plug
  discovery: false
  broker: mqtt.server
```

### Anything in the logs that might be useful for us?

```txt
ryan@hostname:/$ docker run --rm -v "${PWD}":/config -it esphome/esphome run outdoor.yml
INFO Reading configuration outdoor.yml...
INFO Generating C++ source...
INFO Core config or version changed, cleaning build files...
INFO Compiling app...
Processing outdoorplug (board: d1_mini; framework: arduino; platform: platformio/espressif8266 @ 2.6.3)
------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.3
|-- <ESPAsyncWebServer-esphome> 2.1.0
|   |-- <ESPAsyncTCP-esphome> 1.2.3
|   |-- <Hash> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <DNSServer> 1.1.1
|-- <ESP8266WiFi> 1.0
|-- <AsyncMqttClient-esphome> 0.8.6
|   |-- <ESPAsyncTCP-esphome> 1.2.3
|-- <ArduinoJson-esphomelib> 5.13.3
|-- <ESP8266mDNS> 1.2
Compiling .pioenvs/outdoorplug/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/esp8266/core.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/esp8266/gpio.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/esp8266/preferences.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mdns/mdns_esp32_arduino.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mdns/mdns_esp_idf.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/custom_mqtt_device.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_binary_sensor.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_button.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_client.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_climate.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_component.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_cover.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_fan.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_light.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_number.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_select.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_sensor.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_switch.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/mqtt/mqtt_text_sensor.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/ota/ota_backend_arduino_esp32.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/ota/ota_component.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/socket/bsd_sockets_impl.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/socket/lwip_raw_tcp_impl.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/status_led/status_led.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/switch/automation.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/wifi/wifi_component_esp32_arduino.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/components/wifi/wifi_component_esp_idf.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/application.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/color.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/component.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/controller.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/log.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/outdoorplug/src/esphome/core/util.cpp.o
Compiling .pioenvs/outdoorplug/src/main.cpp.o
Generating LD script .pioenvs/outdoorplug/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/outdoorplug/lib975/ESPAsyncTCP-esphome/AsyncPrinter.cpp.o
Compiling .pioenvs/outdoorplug/lib975/ESPAsyncTCP-esphome/ESPAsyncTCP.cpp.o
Compiling .pioenvs/outdoorplug/lib975/ESPAsyncTCP-esphome/ESPAsyncTCPbuffer.cpp.o
Compiling .pioenvs/outdoorplug/lib975/ESPAsyncTCP-esphome/SyncClient.cpp.o
Compiling .pioenvs/outdoorplug/lib975/ESPAsyncTCP-esphome/tcp_axtls.c.o
Compiling .pioenvs/outdoorplug/lib181/Hash/Hash.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Archiving .pioenvs/outdoorplug/lib181/libHash.a
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiGratuitous.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Archiving .pioenvs/outdoorplug/lib975/libESPAsyncTCP-esphome.a
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiServerSecureAxTLS.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pioenvs/outdoorplug/libfb1/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/AsyncEventSource.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/AsyncWebSocket.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/WebAuthentication.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/WebHandlers.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/WebRequest.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/WebResponses.cpp.o
Compiling .pioenvs/outdoorplug/libc74/ESPAsyncWebServer-esphome/WebServer.cpp.o
Compiling .pioenvs/outdoorplug/lib070/DNSServer/DNSServer.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/ConnAckPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PingRespPacket.cpp.o
Archiving .pioenvs/outdoorplug/libfb1/libESP8266WiFi.a
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PubAckPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PubCompPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PubRecPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PubRelPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/PublishPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/SubAckPacket.cpp.o
Compiling .pioenvs/outdoorplug/lib813/AsyncMqttClient-esphome/AsyncMqttClient/Packets/UnsubAckPacket.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/ESP8266mDNS.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/ESP8266mDNS_Legacy.cpp.o
Archiving .pioenvs/outdoorplug/lib070/libDNSServer.a
Archiving .pioenvs/outdoorplug/libc74/libESPAsyncWebServer-esphome.a
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/LEAmDNS.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/LEAmDNS_Control.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/LEAmDNS_Helpers.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/LEAmDNS_Structs.cpp.o
Compiling .pioenvs/outdoorplug/libfb2/ESP8266mDNS/LEAmDNS_Transfer.cpp.o
Archiving .pioenvs/outdoorplug/libFrameworkArduinoVariant.a
Compiling .pioenvs/outdoorplug/FrameworkArduino/Crypto.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Esp-frag.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Esp-version.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Esp.cpp.o
Archiving .pioenvs/outdoorplug/lib813/libAsyncMqttClient-esphome.a
Compiling .pioenvs/outdoorplug/FrameworkArduino/FS.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/FSnoop.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/FunctionalInterrupt.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/HardwareSerial.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/IPAddress.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/MD5Builder.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Print.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Schedule.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/StackThunk.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Stream.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/StreamString.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/Tone.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/TypeConversion.cpp.o
Archiving .pioenvs/outdoorplug/libfb2/libESP8266mDNS.a
Compiling .pioenvs/outdoorplug/FrameworkArduino/Updater.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/WMath.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/WString.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/abi.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/base64.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/cbuf.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/cont.S.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/cont_util.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_eboot_command.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_features.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_flash_quirks.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_flash_utils.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_i2s.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_main.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_noniso.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_phy.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_postmortem.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_si2c.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_sigma_delta.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_spi_utils.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_timer.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_waveform.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring_analog.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring_digital.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring_pulse.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring_pwm.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/core_esp8266_wiring_shift.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/crc32.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/debug.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/flash_hal.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/gdb_hooks.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/heap.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/libb64/cdecode.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/libb64/cencode.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/libc_replacements.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/sntp-lwip2.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs/spiffs_cache.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs/spiffs_check.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs/spiffs_gc.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs/spiffs_hydrogen.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs/spiffs_nucleus.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/spiffs_api.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/sqrt32.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/time.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/uart.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/umm_malloc/umm_info.c.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/umm_malloc/umm_integrity.c.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/umm_malloc/umm_local.c.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/umm_malloc/umm_malloc.cpp.o
Compiling .pioenvs/outdoorplug/FrameworkArduino/umm_malloc/umm_poison.c.o
Archiving .pioenvs/outdoorplug/libFrameworkArduino.a
Linking .pioenvs/outdoorplug/firmware.elf
xtensa-lx106-elf-g++: fatal error: -fuse-linker-plugin, but liblto_plugin.so not found
compilation terminated.
*** [.pioenvs/outdoorplug/firmware.elf] Error 1
============================================= [FAILED] Took 72.60 seconds =============================================
```

### Additional information

It is also worth mentioning that this same Docker image successfully compiled ESP32-based images without issue, and with a small (required) modification to the yaml from:

```yaml
esphome:
  name: outdoorplug

esp8266:
  board: d1_mini
```

to:

```yaml
esphome:
  name: outdoorplug
  platform: esp8266
  board: d1_mini
```

Docker image tagged 2021.9.3 **successfully** compiles the above (later images still fail).

**Comment by OttoWinter** [comment]
Try removing the `.esphome` folder in the current directory ($PWD) and trying again. If that does not work, please say what sort of filesystem (ext4, ntfs) the current directory is in.

**Comment by djryanj** [comment]
@OttoWinter - deleting the `.esphome` folder was successful. It compiles fine.

Even though that was successful, it may be worth noting that my environment technically is Docker for Windows and the filesystem is NTFS, but mounted into WSL which is where I was running Docker from. Perhaps a source of the issue.

I believe I also tried running it from PowerShell, but the result is the same, however, I cannot replicate that now since I deleted `.esphome`.

Thanks for your help.
